invoice/src/main/java/org/killbill/billing/invoice/InvoicePluginDispatcher.java

```diff
@@ -48,30 +55,116 @@
-                                                                                                                    InvoiceItemType.ITEM_ADJ,
-                                                                                                                    InvoiceItemType.CREDIT_ADJ,
-                                                                                                                    InvoiceItemType.TAX);
+                                                                                                                   InvoiceItemType.ITEM_ADJ,
+                                                                                                                   InvoiceItemType.CREDIT_ADJ,
+                                                                                                                   InvoiceItemType.TAX);
 
     private final OSGIServiceRegistration<InvoicePluginApi> pluginRegistry;
     private final InvoiceConfig invoiceConfig;
 
-
     @Inject
     public InvoicePluginDispatcher(final OSGIServiceRegistration<InvoicePluginApi> pluginRegistry,
                                    final InvoiceConfig invoiceConfig) {
         this.pluginRegistry = pluginRegistry;
         this.invoiceConfig = invoiceConfig;
     }
 
+    public DateTime priorCall(final LocalDate targetDate, final List<Invoice> existingInvoices, final boolean isDryRun, final boolean isRescheduled, final CallContext callContext, final InternalTenantContext internalTenantContext) throws InvoiceApiException {
+        log.debug("Invoking invoice plugins priorCall: targetDate='{}', isDryRun='{}', isRescheduled='{}'", targetDate, isDryRun, isRescheduled);
+        final Map<String, InvoicePluginApi> invoicePlugins = getInvoicePlugins(internalTenantContext);
+        if (invoicePlugins.isEmpty()) {
+            return null;
+        }
+
+        DateTime earliestRescheduleDate = null;
+        final InvoiceContext invoiceContext = new DefaultInvoiceContext(targetDate, null, existingInvoices, isDryRun, isRescheduled, callContext);
+        for (final String invoicePluginName : invoicePlugins.keySet()) {
+            final PriorInvoiceResult priorInvoiceResult = invoicePlugins.get(invoicePluginName).priorCall(invoiceContext, ImmutableList.<PluginProperty>of());
+            log.debug("Invoice plugin {} returned priorInvoiceResult='{}'", invoicePluginName, priorInvoiceResult);
+            if (priorInvoiceResult == null) {
+                // Naughty plugin...
+                continue;
+            }
+
+            if (priorInvoiceResult.getRescheduleDate() != null &&
+                (earliestRescheduleDate == null || earliestRescheduleDate.compareTo(priorInvoiceResult.getRescheduleDate()) > 0)) {
+                earliestRescheduleDate = priorInvoiceResult.getRescheduleDate();
+                log.info("Invoice plugin {} rescheduled invoice generation to {} for targetDate {}", invoicePluginName, earliestRescheduleDate, targetDate);
+            }
```

Comment:
Yeah, technically, there is some semantics discussions to have. In practice, if different plugins return different dates, finding the right semantics may be tricky...